packages/web-sdk/src/instrumentations/session/sessionHandler.ts

```diff
@@ -0,0 +1,218 @@
+  return inactivityPeriodValid;
+}
+
+interface PersistentUserSessionUpdater {
+  handleUpdate: () => void;
+  init: () => void;
+}
+
+export function persistentUserSessionsUpdater(initialSessionId?: string): PersistentUserSessionUpdater {
+  const throttledSessionUpdate = throttle(() => {
+    const sessionFromLocalStorage = fetchUserSession();
+
+    if (isUserSessionValid(sessionFromLocalStorage)) {
+      storeUserSession({ ...sessionFromLocalStorage!, lastActivity: dateNow() });
+    } else {
+      let newSession = createUserSessionObject(); // create local srorage session
+      newSession = {
+        ...newSession,
+        sessionMeta: {
+          id: newSession.sessionId,
+          attributes: {
+            ...(faro.metas.value.session?.attributes ?? {}),
+            ...(sessionFromLocalStorage != null ? { previousSession: sessionFromLocalStorage.sessionId } : {}),
+          },
+        },
+      };
+
+      storeUserSession(newSession);
+
+      faro.api?.setSession(newSession.sessionMeta);
+      faro.config.experimentalSessions?.onSessionChange?.(
+        sessionFromLocalStorage?.sessionMeta ?? null,
+        newSession.sessionMeta!
+      );
+    }
+  }, STORAGE_UPDATE_DELAY);
+
+  function init() {
+    storeUserSession(createUserSessionObject(initialSessionId));
+
+    document.addEventListener('visibilitychange', () => {
+      if (document.visibilityState === 'visible') {
+        throttledSessionUpdate();
+      }
+    });
+
+    window.addEventListener('storage', (event: StorageEvent) => {
+      if (event.key !== STORAGE_KEY) {
+        return;
+      }
+
+      const newSession: FaroUserSession = JSON.parse(event.newValue ?? '');
+      const previousSession: FaroUserSession = JSON.parse(event.oldValue ?? '');
+
+      if (newSession.sessionId !== previousSession.sessionId) {
+        faro.api?.setSession(newSession.sessionMeta);
+      }
+    });
+
+    // Users can call the setSession() method, so we need to sync this with the local storage session
+    faro.metas.addListener(function syncSessionIfChangedExternally(meta: Meta) {
+      const session = meta.session;
+      const sessionFromLocalStorage = fetchUserSession();
+
+      if (session && session.id !== sessionFromLocalStorage?.sessionId) {
+        const userSession = {
+          ...createUserSessionObject(session.id),
+          sessionMeta: {
+            id: session.id,
+            attributes: {
+              ...(sessionFromLocalStorage?.sessionMeta?.attributes ?? {}),
+              ...session.attributes,
+              ...(sessionFromLocalStorage ? { previousSession: sessionFromLocalStorage.sessionId } : {}),
+            },
+          },
+        };
+
+        storeUserSession(userSession);
+        faro.api.setSession(userSession.sessionMeta);
+      }
+    });
+  }
+
+  return {
+    handleUpdate: throttledSessionUpdate,
+    init,
+  };
+}
+
+interface InMemoryUserSessionsUpdater {
```

Comment:
This interface is the same as the PeristentSessionUpdater right? Maybe merge them into one?

Reply:
Oh good catch, yes this is correct. Will merge them.